src/modules/python/pbs/v1/_pmi_utils.py

```diff
@@ -0,0 +1,158 @@
+#   
+#  You should have received a copy of the GNU Affero General Public License along 
+#  with this program.  If not, see <http://www.gnu.org/licenses/>.
+#   
+#  Commercial License Information: 
+#  
+#  The PBS Pro software is licensed under the terms of the GNU Affero General 
+#  Public License agreement ("AGPL"), except where a separate commercial license 
+#  agreement for PBS Pro version 14 or later has been executed in writing with Altair.
+#   
+#  Altair’s dual-license business model allows companies, individuals, and 
+#  organizations to create proprietary derivative works of PBS Pro and distribute 
+#  them - whether embedded or bundled with other software - under a commercial 
+#  license agreement.
+#  
+#  Use of Altair’s trademarks, including but not limited to "PBS™", 
+#  "PBS Professional®", and "PBS Pro™" and Altair’s logos is subject to Altair's 
+#  trademark licensing policies.
+ *
+ */
+"""
+__doc__ = """
+$Id$
+Utility power functions.
+"""
+
+import pbs
+import os
+import sys
+
+
+def _pbs_conf(confvar):
+    # Return the value of a setting in the pbs.conf file if it exists.
+    # Save the values in a global dictionary for future use.
+
+    if confvar in os.environ:
+        return os.environ[confvar]
+
+    global pmi_pbsconf
+    if "pmi_pbsconf" not in globals():
+        pmi_pbsconf = dict()
+        cfile = "PBS_CONF_FILE"
+        if cfile in os.environ:
+            pbsconf = os.environ[cfile]
+        else:
+            pbsconf = "/etc/pbs.conf"
+
+        try:
+            fp = open(pbsconf)
+        except:
+            pbs.logmsg(pbs.DEBUG, "%s: conf file not found" % pbsconf)
+            pass
+        else:
+            for line in fp:
+                line = line.strip()
+                # ignore empty lines or those beginning with '#'
+                if line == "" or line[0] == "#":
+                    continue
+                var, eq, val = line.partition('=')
+                if val == "":
+                    continue
+                pmi_pbsconf[var] = val
+            fp.close()
+    if confvar in pmi_pbsconf:
+        return pmi_pbsconf[confvar]
+    else:
+        return None
+
+
+def _is_node_provisionable():
+    """
+    Check if the local machine is running either the pbs_server or
+    pbs_sched.  If either of these are running, provisioning
+    must not be automatically enabled.
+    """
+    serv = _pbs_conf("PBS_START_SERVER")
+    if serv is not None and serv == "1":
+        return False
+
+    sched = _pbs_conf("PBS_START_SCHED")
+    if sched is not None and sched == "1":
+        return False
+    return True
+
+
+def _get_hosts(job):
+    """
+    Form a list of unique short hostnames from a pbs job.
+    The short names are used even if the hostnames from exec_host2
+    are FQDNs because SGIMC seems to have a bug in that it will
+    not accept FQDNs.
+    """
+    hosts = str(job.exec_host2)
+    pbs_nodes = sorted(set([x.partition(':')[0].partition('.')[0]
```

Comment:
Sorted list looks good on logs and stats (SGI can stat these nodeset and see what all nodes are part of it) and since we pass this list to external api (here SGI api Nodesetcreate) I would have it sorted before sending for optimal usage.